prov/ucx/src/ucx.h

```diff
@@ -0,0 +1,320 @@
+	struct dlist_entry rkey_list;
+	uint64_t signature;
+	size_t pkey_size;
+	uint8_t pkey[0];
+};
+
+/* unpacked raw key, one per peer per MR */
+struct ucx_mr_rkey {
+	struct {
+		fi_addr_t owner_addr;
+		uint64_t key;
+	} id;
+	ucp_rkey_h rkey;
+	UT_hash_handle hh;
+	struct dlist_entry entry;
+};
+
+struct ucx_domain {
+	struct util_domain u_domain;
+	ucp_context_h context;
+	struct ucx_mr_rkey *remote_keys;
+};
+
+struct ucx_ep {
+	struct util_ep ep;
+	struct ucx_av *av; /*until AV is not implemented via utils*/
+	ucp_worker_h worker;
+	short service;
+	void *addr;
+	size_t addr_len;
+	struct dlist_entry claimed_list;
+	struct {
+		size_t mrecv_min_size;
+	} ep_opts;
+	struct dlist_entry mctx_freelist;
+	struct dlist_entry mctx_repost;
+};
+
+struct ucx_ave {
+	ucp_ep_h uep;
+	void *addr;
+};
+
+struct ucx_avblock {
+	struct ucx_avblock *next;
+	char payload[0];
+};
+
+struct ucx_av {
+	struct fid_av av;
+	struct ucx_domain *domain;
+	struct ucx_ep *ep;
+	struct util_eq *eq;
+	int type;
+	int async;
+	size_t count;
+	size_t addr_len;
+	struct ucx_avblock *addr_blocks;
+	struct ucx_avblock *ep_block;
+};
+
+#define UCX_GET_UCP_EP(EP, ADDR) (((struct ucx_ave*)(ADDR))->uep)
+
+typedef enum ucx_comm_mode {
+	UCX_MSG,
+	UCX_TAGGED,
+} ucx_comm_mode_t;
+
+typedef enum ucx_req_type {
+	UCX_FI_REQ_UNINITIALIZED = 0,
+	UCX_FI_REQ_READ = 0x01,
+	UCX_FI_REQ_WRITE = 0x02,
+	UCX_FI_REQ_MULTIRECV = 0xAA,
+	UCX_FI_REQ_MULTIRECV_UNEXP = 0xAB,
+	UCX_FI_REQ_REGULAR = 0xFD,
+	UCX_FI_REQ_UNEXPECTED_ERR = 0xFE,
+	UCX_FI_REQ_UNEXPECTED = 0xFF,
+} ucx_req_type_t;
+
+struct ucx_mrecv_ctx {
+	struct dlist_entry list;
+
+	/* user buffer state */
+	ssize_t remain;
+	void *head;
+
+	/* read-only repost/completion info */
+	uint64_t tag;
+	uint64_t tag_mask;
+	uint64_t flags;
+	void *context;
+};
+
+struct ucx_mrecv_request {
+	ucx_req_type_t type;
+	struct util_cq* cq;
+	struct ucx_ep* ep;
+
+	/* completion status*/
+	ssize_t last_recvd;
+	ucs_status_t status;
+
+	struct ucx_mrecv_ctx *mrecv_ctx;
+};
+
+struct ucx_request {
+	ucx_req_type_t type;
+	struct util_cq* cq;
+	struct ucx_ep* ep;
+
+	/* completion status*/
+	ssize_t posted_size;
+	ucs_status_t status;
+
+	union {
+		struct fi_cq_tagged_entry tagged;
+		struct fi_cq_data_entry data;
+	} completion;
+};
+
+static inline void ucx_req_release(struct ucx_request *req) {
+	req->type = UCX_FI_REQ_UNINITIALIZED;
+	ucp_request_free(req);
+}
+
+#define UCX_EP_MSG_TAG (~0ULL)
+#define UCX_AUX_TAG_MASK (~0ULL)
+
+OFI_DECLARE_CIRQUE(struct fi_cq_tagged_entry, ucx_comp_cirq);
+
+extern int ucx_errcode_translation_table[];
+#define UCX_TRANSLATE_ERRCODE(X) ucx_errcode_translation_table[(-X)+1]
+extern struct fi_provider ucx_prov;
+extern struct ucx_global_descriptor ucx_descriptor;
+extern struct util_prov ucx_util_prov;
+
+extern struct fi_ops_cm ucx_cm_ops;
+extern struct fi_ops_tagged ucx_tagged_ops;
+extern struct fi_ops_msg ucx_msg_ops;
+extern struct fi_ops_mr ucx_mr_ops;
+extern struct fi_ops_rma ucx_rma_ops;
```

Comment:
The mr ops doesn't need to be external. The other three are defined in ucx_tagged.c, ucx_msg.c and ucx_rma.c, respectively but are used in ucx_ep.c.  Moving the definitions to ucx_ep.c would then require exposing a bunch of more functions as external, which is even less desirable.

Reply:
I was suggesting moving the extern definitions into ucx_ep.c, rather than moving the functions themselves.  The functions are probably in the right files.

Reply:
that makes sense.